Lease

mshitrit opened the pull request:
Adding some basics to common repo, and trying to start set up `lease logic` to be used from it.

mshitrit committed:
backport lease logic from nmo

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit committed:
add Makefile

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit committed:
fix lint issues

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit committed:
refactor holder identity into a param

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit committed:
cancel default nmo lease namespace

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit committed:
add owners file

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit committed:
add dockerized test

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit committed:
trigger unit tests with git action

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit committed:
update gitignore

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit committed:
init module

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit committed:
go mod vendor

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

slintes commented:
WDYT about having a lease manager with a clear interface? Something like (just a quick shot, details might vary):

```
type LeaseManager interface {
	CreateOrGetLease(ctx context.Context, node *corev1.Node, duration time.Duration, holderIdentity string, namespace string) (*coordv1.Lease, bool, error)
	UpdateLease(ctx context.Context, node *corev1.Node, lease *coordv1.Lease, currentTime *metav1.MicroTime, leaseDuration, leaseDeadline time.Duration, holderIdentity string) (bool, error)
	InvalidateLease(ctx context.Context, nodeName string, leaseNamespace string) error
}

type leaseManager struct {
	client.Client
	log logr.Logger
}

// enforce interface implementation
var _ LeaseManager = leaseManager{}

func NewManager(cl client.Client, log logr.Logger) LeaseManager {
	return &leaseManager{
		Client: cl,
		log:    log,
	}
}

func (lm *leaseManager) CreateOrGetLease(...) ... {...}
```

I think every public method should have a Context as first arg, and return an error as last value, even when it's not used today. Would prevent annoying API changes in case we need it later 🙂

mshitrit commented:
>WDYT about having a lease manager with a clear interface?

TBH initially I was just focused on adjusting the same code to a public repo.
I agree that interface, is the proper way to go here, and the overhead is small so I'll defiantly squeeze it into this PR.

mshitrit committed:
- create API to use lease
- remove check for lease support - supported at least since k8s 1.22
- minor lint and deprecated APIs fixes

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit committed:
remove unused param / methods

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit committed:
use context

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit committed:
rename files from lease to manager

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

slintes commented on go.mod:
```diff
@@ -0,0 +1,75 @@
+module github.com/medik8s/common
+
+go 1.19
```
please update to 1.20 (NHC uses 1.20)

slintes commented on Makefile:
```diff
@@ -0,0 +1,37 @@
+# GO_VERSION refers to the version of Golang to be downloaded when running dockerized version
+GO_VERSION = 1.19
```
1.20

slintes commented on Makefile:
```diff
@@ -0,0 +1,37 @@
+export DOCKER_GO=docker run --rm -v $$(pwd):/home/go/src/github.com/medik8s/common \
+	-u $$(id -u) -w /home/go/src/github.com/medik8s/common \
+	-e "GOPATH=/go" -e "GOFLAGS=-mod=vendor" -e "XDG_CACHE_HOME=/tmp/.cache" \
+	-e "VERSION=$(VERSION)" -e "IMAGE_REGISTRY=$(IMAGE_REGISTRY)" \
```
VERSIOn and IMAGE_REGISTRY (and DEFAULT_VERSION) aren't needed

slintes commented on Makefile:
```diff
@@ -0,0 +1,37 @@
+
+
+.PHONY: test
+test: ## Run tests.
```
I would like to have some more basic checks here: go mod tidy and vendor, fix-imports, goimports, vet, verify modified files

slintes commented on pkg/lease/manager.go:
```diff
@@ -0,0 +1,205 @@
+)
+
+type Manager interface {
+	CreateOrGetLease(ctx context.Context, node *corev1.Node, duration time.Duration, holderIdentity string, namespace string) (*coordv1.Lease, bool, error)
```
what about making this more generic and not using Node but client.Object? (applies to all places where Node is used)

mshitrit replied:
IIUC this will add a bit of complication to the implementation (i.e checking that the element is a Node, and error handling).
is there any particular reason we want to do it ? (i.e do we plan to supported lease for non Nodes ?)

slintes replied:
> checking that the element is a Node hand error handling

I do not understand?

> is there any particular reason

I just think that there isn't a reason to limit lease funtionality to Nodes

mshitrit replied:
>I do not understand?

Fixed typo in my original comment.

In a nutshell I've wrongly assumed that lease functionality is only relevant for Nodes, so I've thought that in the implementation we need something like that:
```
if node, ok  := object.(*corev1.Node); !ok{
  //handle error
}
```

slintes replied:
> lease functionality is only relevant for Nodes

It's the only usecase for now, and I'm not aware of any other yet. But code wise it doesn't make a big difference if we use Node or Object as argument, so IMHO it doesn't make sense to limit this library to Nodes. So we also don't need that error handling :)

slintes commented on pkg/lease/manager.go:
```diff
@@ -0,0 +1,205 @@
+
+	lease := &coordv1.Lease{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            node.ObjectMeta.Name,
```
would make sense to include e.g kind of object, if we use client.Object instead of Node

slintes replied:
still relevant: since this can be used for more things than just nodes, we can run into conflicts when creating leases of different resources but same name. So my suggestion is to prefix the name with the kind of the resource

slintes replied:
still unresolved

mshitrit replied:
Hmm maybe I misunderstood your intention, isn't [this](https://github.com/medik8s/common/pull/1/commits/310da66a49ddb25e6fcc6043f3992a2268b2c8e6)
what you meant by that comment ?

slintes commented on pkg/lease/manager.go:
```diff
@@ -0,0 +1,205 @@
+)
+
+type Manager interface {
+	CreateOrGetLease(ctx context.Context, node *corev1.Node, duration time.Duration, holderIdentity string, namespace string) (*coordv1.Lease, bool, error)
```
please add comments

slintes replied:
still relevant
what is the method doing, what are the parameters, and most important (because they are unnamed): what's the meaning of the return values, especially bools?

slintes commented on pkg/lease/manager.go:
```diff
@@ -0,0 +1,205 @@
+
+type Manager interface {
+	CreateOrGetLease(ctx context.Context, node *corev1.Node, duration time.Duration, holderIdentity string, namespace string) (*coordv1.Lease, bool, error)
+	UpdateLease(ctx context.Context, node *corev1.Node, lease *coordv1.Lease, currentTime *metav1.MicroTime, leaseDuration, leaseDeadline time.Duration, holderIdentity string) (bool, error)
```
IMHO for test purposes `currentTime` on the internal method is useful... but in the API I'm not sure. I would skip it and call internal with now. WDYT?

slintes replied:
ping?

mshitrit replied:
Indeed I agree

slintes commented on pkg/lease/manager.go:
```diff
@@ -0,0 +1,205 @@
+
+type Manager interface {
+	CreateOrGetLease(ctx context.Context, node *corev1.Node, duration time.Duration, holderIdentity string, namespace string) (*coordv1.Lease, bool, error)
+	UpdateLease(ctx context.Context, node *corev1.Node, lease *coordv1.Lease, currentTime *metav1.MicroTime, leaseDuration, leaseDeadline time.Duration, holderIdentity string) (bool, error)
```
I don't remember why we had the `leaseDeadline` arg... doesn't make much sense to me. Isn't leaseDuration good enough? Just always update the lease with that duration? I don't see a usecase for "only update lease if it expires before deadline". WDYT?

slintes replied:
ping

mshitrit replied:
I think it's meant to allow more flexibility.

For example we can specify we'd like to reset the lease to expire 30 minutes  from now (leaseDuration) in case it's going to expire in the next hour (dealine).

In case we only use leaseDuration I don't think this sort of request can be accomplished.

slintes replied:
> reset the lease to expire 30 minutes from now (leaseDuration) in case it's going to expire in the next hour (dealine)

yes, but as I wrote in my comment above already, what's the usecase? Why would someone want to set lease `duration` depending on a `deadline`? Either I need the lease for `duration` or not... 🤔 IMHO this adding unneeded complexity to the API and to the implementation...

slintes commented on pkg/lease/manager.go:
```diff
@@ -0,0 +1,205 @@
+
+		}
+	} else {
+		// can't update the lease if it is currently valid.
```
nit: "can't take over the lease" would be clearer IMHO

slintes replied:
ping

mshitrit replied:
Yep, I agree

slintes commented on pkg/lease/manager.go:
```diff
@@ -0,0 +1,205 @@
+
+func makeExpectedOwnerOfLease(node *corev1.Node) *metav1.OwnerReference {
+	return &metav1.OwnerReference{
+		APIVersion: corev1.SchemeGroupVersion.WithKind("Node").Version,
```
get kind from arg

slintes commented on pkg/lease/manager.go:
```diff
@@ -0,0 +1,205 @@
+		lease.Spec.RenewTime = currentTime
+		if err := l.Client.Update(ctx, lease); err != nil {
+			log.Errorf("Failed to update the lease. node %s error: %v", node.Name, err)
+			return updateAlreadyOwnedLease, err
```
the meaning / use case of the 1st return value is mysterious to me :D

mshitrit replied:
Hmm IIUC the boolean flag means whether the update was successful or not, however since 
this is mostly copy pasted from [here](https://github.com/medik8s/node-maintenance-operator/blame/main/controllers/lease.go#L169) you might want to ask the author of the original for a better answer  (I think he is called Marc Suliter  :wink:)

slintes replied:
I always question code from the past, no matter who wrote it, including me 🙂 

> whether the update was successful or not

that's what we have the error for, not? 🤔

mshitrit replied:
Looking into this further I think that the bool is an indication whether prior to the update the lease was already held by holderIdentity (true) or not (false).

slintes commented on pkg/lease/manager.go:
```diff
@@ -0,0 +1,205 @@
+type Manager interface {
+	CreateOrGetLease(ctx context.Context, node *corev1.Node, duration time.Duration, holderIdentity string, namespace string) (*coordv1.Lease, bool, error)
+	UpdateLease(ctx context.Context, node *corev1.Node, lease *coordv1.Lease, currentTime *metav1.MicroTime, leaseDuration, leaseDeadline time.Duration, holderIdentity string) (bool, error)
+	InvalidateLease(ctx context.Context, nodeName string, leaseNamespace string) error
```
this needs to have the object we create the lease for as arg, not the just name, like in the create method IMHO (at least when we go the generic way)

mshitrit replied:
Good chance that I've missed something since I ended up just using the name.
Let me know if you can point me in the correct direction.

slintes replied:
related to the comment about the name of the lease (adding the object kind for uniqueness): when we delete the lease, we also need both kind and name

mshitrit replied:
Putting aside the question whether InvalidateLease should delete or update the Lease, IIUC kind will be needed in any case either for Get or Delete, right ?

slintes replied:
yes, that's why I proposed to use client.Object here like in the other methods

slintes commented on pkg/lease/manager.go:
```diff
@@ -0,0 +1,205 @@
+}
+
+func (l *manager) invalidateLease(ctx context.Context, nodeName string, leaseNamespace string) error {
+	log.Info("Lease object supported, invalidating lease")
```
skip "Lease object supported"

slintes commented on pkg/lease/manager.go:
```diff
@@ -0,0 +1,205 @@
+	lease.Spec.RenewTime = nil
+	lease.Spec.LeaseTransitions = nil
+
+	if err := l.Client.Update(ctx, lease); err != nil {
```
I'm wondering why we don't delete the lease...?

mshitrit replied:
Good question, no idea.
My best guess is that this will eventually trigger the lease removal and maybe has a benefit of trying to catch an update 409 conflict :thinking: 

However I'm not sure it'll be wise to make this kind of change of functionality in this PR (I see it as more of just exporting current lease functionality to common with a bit of a "face lift"  and this sort of change feels more like an "open heart surgery" :slightly_smiling_face: )

slintes replied:
> this will eventually trigger the lease removal 

no one will delete the lease, if not we. There isn't something like a lease controller which deletes expired leases (to my best knowledge)

> However I'm not sure it'll be wise to make this kind of change of functionality in this PR (I see it as more of just exporting current lease functionality to common with a bit of a "face lift" and this sort of change feels more like an "open heart surgery" slightly_smiling_face )

Well, I see it otherwise. We are creating an API here, which will be used by multiple projects, and I don't think we want to update the API every few days or weeks when we notice that the API isn't ideal. I'd prefer to create a as-clean-as-possible API right away, and not a copy/paste of code from years ago.
We have the awsome chance to create the API while already having tests on a project which is going to use it. That's not a open heart surgery IMHO 🙂 
That's why I question the public methods very much, but also some impementation details. No matter who wrote the old code.

mshitrit replied:
>I'd prefer to create a as-clean-as-possible API right away, and not a copy/paste of code from years ago.

A full API examination/refactor wasn't my original intention, but since in any case we're almost there I'm happy to be onboard :sunglasses: 

- I think that the upside of updating the lease is saving resources in case of the lease is often used  (I would assume that constant delete / creation would take more resources than update) and a  faster acquirement if the lease.
- The downside is cluttering the etcd with unused resources (in case the lease is rarely used).

Let me know if you think otherwise, or have other considerations which I've missed and which approach do you prefer.

slintes replied:
IMHO the downside effect is bigger than the upside 🙂

mshitrit committed:
update go version to 1.20

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit committed:
removed unused params from Makefile

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit committed:
Adding more verifications

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit committed:
sort imports

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit committed:
switch node to client.Object

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit committed:
minor logging improvement

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

slintes commented:
Thanks for adding more checks to the Makefile. ~~What about adding a github workflow to call them? 🙂~~

edit: sorry, I saw there is, I was confused by not seeing it executed 🤔

mshitrit committed:
adding kind & APIVersion to lease creation

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit committed:
Documenting the API methods

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit committed:
Remove nmo as hard coded owner in log

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit committed:
Remove now time from UpdateLease API

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit committed:
improving comment readability

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit commented:
> Thanks for adding more checks to the Makefile. ~What about adding a github workflow to call them? slightly_smiling_face~
> 
> edit: sorry, I saw there is, I was confused by not seeing it executed thinking

Yep I was wondering about that too, I've figured that it has to do with origin been completely empty now and that we'll have the checks running in the second PR - but it could be something else I'm missing :man_shrugging:

mshitrit committed:
Adding namespace and holderIdentity to manager
Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit committed:
refactor InvalidateLease

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit committed:
refactor UpdateLease

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit committed:
Add GetLease API

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit committed:
Minor logs

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit committed:
increase UT coverage

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

slintes commented on pkg/lease/manager.go:
```diff
@@ -0,0 +1,235 @@
+	log            logr.Logger
+}
+
+func (l *manager) RequestLease(ctx context.Context, obj client.Object, leaseDuration time.Duration) error {
```
What is the purspose of this pattern,letting the public method just call a private method with the same signature? 🤔

mshitrit replied:
Basically it creates a separation between the API (public method) which needs to follow up the interface signature and the implementation (the private method) which does not.
I think it was much more cleared when they actually had different signatures (when `now` was passed as variable to the implementation) but I still see value in this separation both conceptually and because IMO it better protects the signature of the API.

slintes replied:
> separation between the API (public method) which needs to follow up the interface signature 

But the interface is the API, not? While the methods here are public, their receiver (the `manager` struct) isn't. So in the end we are private here already.

> it better protects the signature of the API.

 I don't see how this protects the API. As long as the interface doesn't change, the API is safe.


> I still see value in this separation [...] conceptually

I can't remember seeing this pattern before, is this best practice? Maybe I'm just missing something 🤔

mshitrit replied:
>But the interface is the API, not? 

[This](https://github.com/medik8s/common/pull/1/files#diff-a4d4edb8162b6478f493479ca04ccd20064f12c52817174cba8eff2b7738f978R20-R29) is what I consider the interface (in that  context)

>While the methods here are public, their receiver (the manager struct) isn't. So in the end we are private here already.

You are correct in the sense that it can't be directly accessed from a different packages, but that was not my point, my point was to create distinction between `RequestLease` which signature shouldn't be changed to `requestLease` which is more flexible.

>I don't see how this protects the API. As long as the interface doesn't change, the API is safe.

It is true, however it's fairly easy to refactor `RequestLease` so it'll change the API as well - most IDEs will do it for you.
I believe that such mistake is easier to make when RequestLease is both the API and the implementation, rather than the use case where we have such separation. 

>I can't remember seeing this pattern before, is this best practice? 

No idea, at least it's not something I'm aware of, I think maybe it has to do with the fact that IMO in go the coupling between implementation and it's interface is looser than in some other languages (Java comes to mind in that aspect).
Here we only relay on return type of  the initialization being the interface, if we change `Manager` to `*manager`  we loose all coupling to the interface.

All that being said I don't see any of it as a major issue (although I do like a good discussion) , so if you feel this pattern creates more confusion than value I wouldn't mind removing the private methods.

slintes replied:
>  it's fairly easy to refactor RequestLease so it'll change the API as well - most IDEs will do it for you.

I still doubt the value over the confusion, but this is a point at least. So I'm fine with keeping it.

slintes commented on pkg/lease/manager.go:
```diff
@@ -0,0 +1,235 @@
+	nName := apitypes.NamespacedName{Namespace: l.namespace, Name: obj.GetName()}
+	lease := &coordv1.Lease{}
+
+	getOption := &metav1.GetOptions{
```
What do we need this for? 🤔

mshitrit replied:
Based on your question I think I'm probably missing something :sweat_smile: , but I'll try to explain my line of thinking:

1. when we create the lease [here](https://github.com/medik8s/common/pull/1/files#diff-a4d4edb8162b6478f493479ca04ccd20064f12c52817174cba8eff2b7738f978R64-R93) we add to the TypeMeta the Kind & Version because we'd like to be able to distinct between a lease called `abc` where client.Object is a `node ` and a lease called `abc` where client.Object is a `pod`
2. So when we retrieve the lease here we'd also like to make the same distinction to make sure we get the proper one.

slintes replied:
oooohhhhh..... I missed that in the create function.

`TypeMeta` is about the type which is being created or read, so the Lease! Not the type of the object we create the Lease for. Basically you are using the Go `Lease` type now, but override the type internally with something different. Very interesting that it doesn't break the unit tests...

> we'd like to be able to distinct between a lease called abc where client.Object is a node  and a lease called abc where client.Object is a pod

My idea was to add the kind to the name of the lease as prefix. Using a label could also be an idea.

mshitrit replied:
Hmm I do have a question:
Do we want to add the `Kind / Version` of client.Object in order to make sure lease key is unique ?
In case the answer is yes, I think your suggestion of adding it as a prefix to the name makes sense, but I don't think it'll apply for using a label.

Or maybe we just want to make sure we are getting the correct lease when fetching ? 
In that case maybe we can just use the OwnerReferences which already has the Kind/Version (and than we don't need to add labels or manipulate the name).

WDYT ?

slintes replied:
We need a unique lease name over all object kinds, so we can have a lease for a Pod named `foo` and a Node named `foo` at the same time. So yes, adding the kind (without version) to the name it is. Labels don't fix issues indeed, my bad.

mshitrit committed:
update UT to cover creation of multiple leases which owner has the same name

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit committed:
Add lease owner kind as prefix to lease name

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit committed:
removed unused test const

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

slintes commented on pkg/lease/manager.go:
```diff
@@ -0,0 +1,228 @@
+}
+
+func generateLeaseName(obj client.Object) string {
+	return fmt.Sprintf("%s_%s", obj.GetObjectKind().GroupVersionKind().Kind, obj.GetName())
```
underscores are not allowed for names, please use minus sign: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-subdomain-names

slintes reviewed (changes requested):
One small change needed, otherwise lgtm. Looking forward to see the NMO PR updated to use the cleaned up API :)

mshitrit committed:
replace `_` in name with `-`

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>

mshitrit committed:
support fetching Kind/version for node/pod if not found

Signed-off-by: Michael Shitrit <mshitrit@redhat.com>
